File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Queue/Samples/QueueSample.java
package Queue.Samples;
import java.util.ArrayDeque;
import java.util.Deque;

/**
 * Queue for Inventory Processing (FIFO - process in order received).
 */
public class QueueSample {

    record InventoryItem(String brand, String model, int quantity) {}

    static class InventoryQueue {
        private final Deque<InventoryItem> queue = new ArrayDeque<>();

        void enqueue(String brand, String model, int quantity) {
            queue.addLast(new InventoryItem(brand, model, quantity));
        }

        InventoryItem dequeue() {
            if (queue.isEmpty()) {
                return null;
            }
            return queue.removeFirst();
        }

        InventoryItem peek() {
            if (queue.isEmpty()) {
                return null;
            }
            return queue.peekFirst();
        }

        boolean isEmpty() {
            return queue.isEmpty();
        }

        void display() {
            if (queue.isEmpty()) {
                System.out.println("  (queue empty)");
                return;
            }
            int i = 1;
            for (InventoryItem item : queue) {
                System.out.println("  " + i++ + ". " + item.brand() + " " + item.model()
                        + ": " + item.quantity() + " units");
            }
        }
    }

    public static void main(String[] args) {
        InventoryQueue inventoryQueue = new InventoryQueue();

        System.out.println("Enqueueing inventory updates...");
        inventoryQueue.enqueue("Suzuki", "Raider 150", 3);
        inventoryQueue.enqueue("Yamaha", "NMAX 155", 5);
        inventoryQueue.enqueue("Honda", "Click 125", 10);
        System.out.println("Queue (front to back):");
        inventoryQueue.display();

        System.out.println("\nPeek at front (next to process):");
        System.out.println("  " + inventoryQueue.peek());

        System.out.println("\nDequeue (process first item)...");
        InventoryItem processed = inventoryQueue.dequeue();
        System.out.println("  Processed: " + processed);
        System.out.println("Queue after dequeue:");
        inventoryQueue.display();

        System.out.println("\nDequeue next item...");
        processed = inventoryQueue.dequeue();
        System.out.println("  Processed: " + processed);
        System.out.println("Queue after dequeue:");
        inventoryQueue.display();
    }
}

File: 2nd Year/Term 2/Data Structures & Algorithms/MS1/Old files (mistake)/Queue/Samples/queuesample.py
from collections import deque

# Queue for inventory processing (FIFO - process in order received)


class InventoryQueue:
    def __init__(self):
        self.queue = deque()

    # Enqueue: add an inventory update to the back of the queue
    def enqueue(self, brand, model, quantity):
        self.queue.append({
            "brand": brand,
            "model": model,
            "quantity": quantity
        })

    # Dequeue: remove and return the front item (process next in line)
    def dequeue(self):
        if self.is_empty():
            return None
        return self.queue.popleft()

    # Peek: view the front item without removing it
    def peek(self):
        if self.is_empty():
            return None
        return self.queue[0]

    def is_empty(self):
        return len(self.queue) == 0

    def display(self):
        if self.is_empty():
            print("  (queue empty)")
            return
        for i, item in enumerate(self.queue):
            print(f"  {i + 1}. {item['brand']} {item['model']}: {item['quantity']} units")


def main():
    # Queue for inventory processing (MotorPH)
    inventory_queue = InventoryQueue()

    # Enqueue inventory updates
    print("Enqueueing inventory updates...")
    inventory_queue.enqueue("Suzuki", "Raider 150", 3)
    inventory_queue.enqueue("Yamaha", "NMAX 155", 5)
    inventory_queue.enqueue("Honda", "Click 125", 10)
    print("Queue (front to back):")
    inventory_queue.display()

    # Peek at front
    print("\nPeek at front (next to process):")
    front = inventory_queue.peek()
    print(f"  {front}")

    # Dequeue: process first item
    print("\nDequeue (process first item)...")
    processed_item = inventory_queue.dequeue()
    print(f"  Processed: {processed_item}")
    print("Queue after dequeue:")
    inventory_queue.display()

    # Dequeue one more
    print("\nDequeue next item...")
    processed_item = inventory_queue.dequeue()
    print(f"  Processed: {processed_item}")
    print("Queue after dequeue:")
    inventory_queue.display()


if __name__ == "__main__":
    main()
